### PR TITLE
Validate strings from Helm repo before passing to shelljs

### PR DESCRIPTION
**Pull request by itowlson**
Fixes a Repo Explorer issue if user adds a malicious repo and clicks on a malicious chart.

### CODE REVIEW

File: src/helm.documentProvider.ts [hunk 1]
```diff
@@ -78,6 +79,16 @@ export class HelmInspectDocumentProvider implements vscode.TextDocumentContentPr
             } else if (uri.authority === helm.INSPECT_REPO_AUTHORITY) {
                 const id = uri.path.substring(1);
                 const version = uri.query;
+
+                if (!shell.isSafe(id)) {
+                    vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to inspect this chart.`);
```
Comment (lstocchi) [inline src/helm.documentProvider.ts:84]: ${id}

File: src/helm.documentProvider.ts [hunk 2]
```diff
@@ -78,6 +79,16 @@ export class HelmInspectDocumentProvider implements vscode.TextDocumentContentPr
...
+                    return;
+                }
+                if (version && !shell.isSafe(version)) {
+                    vscode.window.showWarningMessage(`Unexpected characters in chart version {version}. Use Helm CLI to inspect this chart.`);
```
Comment (lstocchi) [inline src/helm.documentProvider.ts:88]: ${version}

File: src/helm.exec.ts [hunk 3]
```diff
@@ -403,6 +404,15 @@ export async function helmFetch(helmObject: helmrepoexplorer.HelmObject | undefi
 }
 
 async function helmFetchCore(chartId: string, version: string | undefined): Promise<void> {
+    if (!shell.isSafe(chartId)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to fetch this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:408]: ${chartId}

File: src/helm.exec.ts [hunk 4]
```diff
@@ -403,6 +404,15 @@ export async function helmFetch(helmObject: helmrepoexplorer.HelmObject | undefi
...
+        return;
+    }
+    if (version && !shell.isSafe(version)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart version {version}. Use Helm CLI to fetch this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:412]: ${version}

File: src/helm.exec.ts [hunk 5]
```diff
@@ -432,6 +442,15 @@ export async function helmInstall(kubectl: Kubectl, helmObject: helmrepoexplorer
 }
 
 async function helmInstallCore(kubectl: Kubectl, chartId: string, version: string | undefined): Promise<void> {
+    if (!shell.isSafe(chartId)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to install this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:446]: ${chartId}

File: src/helm.exec.ts [hunk 6]
```diff
@@ -475,6 +494,15 @@ export async function helmDependencies(helmObject: helmrepoexplorer.HelmObject |
 }
 
 async function helmDependenciesLaunchViewer(chartId: string, version: string | undefined): Promise<void> {
+    if (!shell.isSafe(chartId)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to install this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:498]: ${chartId}

File: src/helm.exec.ts [hunk 7]
```diff
@@ -432,6 +442,15 @@ export async function helmInstall(kubectl: Kubectl, helmObject: helmrepoexplorer
 }
 
 async function helmInstallCore(kubectl: Kubectl, chartId: string, version: string | undefined): Promise<void> {
+    if (!shell.isSafe(chartId)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to install this chart.`);
+        return;
+    }
+    if (version && !shell.isSafe(version)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart version {version}. Use Helm CLI to install this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:450]: ${version}

File: src/helm.exec.ts [hunk 8]
```diff
@@ -475,6 +494,15 @@ export async function helmDependencies(helmObject: helmrepoexplorer.HelmObject |
 }
 
 async function helmDependenciesLaunchViewer(chartId: string, version: string | undefined): Promise<void> {
+    if (!shell.isSafe(chartId)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart name {id}. Use Helm CLI to install this chart.`);
+        return;
+    }
+    if (version && !shell.isSafe(version)) {
+        vscode.window.showWarningMessage(`Unexpected characters in chart version {version}. Use Helm CLI to install this chart.`);
```
Comment (lstocchi) [inline src/helm.exec.ts:502]: ${version}

File: src/helm.documentProvider.ts (hunk 1 above)
Comment (itowlson) [reply src/helm.documentProvider.ts:84]: Gah.  Thank you.

**Comment by itowlson** [comment]
Aargh, well that was pretty impressively inept.  Thanks for catching that @lstocchi, sorry for sending you such an obvious blunder, and could you take another look please now?

File: src/helm.documentProvider.ts [hunk 9]
```diff
@@ -78,6 +79,16 @@ export class HelmInspectDocumentProvider implements vscode.TextDocumentContentPr
             } else if (uri.authority === helm.INSPECT_REPO_AUTHORITY) {
                 const id = uri.path.substring(1);
                 const version = uri.query;
+
+                if (!shell.isSafe(id)) {
+                    vscode.window.showWarningMessage(`Unexpected characters in chart name ${id}. Use Helm CLI to inspect this chart.`);
```
Comment (bacongobbler) [inline src/helm.documentProvider.ts:84]: Should this report what the offending characters were in the id?

File: src/helm.documentProvider.ts [hunk 10]
```diff
@@ -78,6 +79,16 @@ export class HelmInspectDocumentProvider implements vscode.TextDocumentContentPr
...
+                    return;
+                }
+                if (version && !shell.isSafe(version)) {
+                    vscode.window.showWarningMessage(`Unexpected characters in chart version ${version}. Use Helm CLI to inspect this chart.`);
```
Comment (bacongobbler) [inline src/helm.documentProvider.ts:88]: Same here. It’s be nice to know why vscode is rejecting to inspect the chart.

File: src/helm.documentProvider.ts (hunk 9 above)
Comment (itowlson) [reply src/helm.documentProvider.ts:84]: That sounds more effort than I'm interested in putting in for a "should never happen" scenario.  _grin_  I can certainly do this but it's a bit fiddly and verbose and I'm not sure it's worth it unless we actually hit this happening in the wild.

(FX: karma spinning up to hit it on day one after release.)

**Pull request merged**